### ISSUE TITLE
Wiki Data Formation Problem

### ISSUE THREAD

**Issue by zheolong** [google-research/bert#341]
I'm checking and converting the wiki data for pre-trainning, just like below:

![11_06_17__01_09_2019](https://user-images.githubusercontent.com/5104916/50874643-d1441900-13ff-11e9-8193-8bd73e7960c9.jpg)

When the process finished, I've got 12792 files, one of these files shows like below:

<img width="1412" alt="2019-01-09 10 58 38" src="https://user-images.githubusercontent.com/5104916/50874796-529bab80-1400-11e9-97cb-b6ca804eaa5c.png">

One line means represents one paragraph, but what we need is one line for one sentence, so am I wrong, or the script is not complete ? Do you have the complete script to do the whole job ?

**Comment by zheolong** [comment]
@dsindex That's what I want. Can u show me an example to run this script?

**Comment by dsindex** [comment]
@zheolong 

https://github.com/dsindex/bert

you can copy 'create_pretraining_data.bash' for that purpose :)

1. download wiki data
  - https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles.xml.bz2

2. extract text
```
$ cd data
$ git clone https://github.com/attardi/wikiextractor
$ cd wikiextractor
$ python WikiExtractor.py -o ../output < ../enwiki-latest-pages-articles.xml
```

3. create pretraining data
  - https://github.com/dsindex/bert/blob/master/create_pretraining_data.bash
```
$ ./create_pretraining_data.bash
```

**Comment by dsindex** [comment]
it read output/AA/wiki_00.txt(for example) and convert to tf record file.
and wiki_xx.txt is made from several wiki pages.
(because there are many doc ids in it).

since create_pretraining_data.py can’t handle very large text file due to out of memory,
we should spilt input text file. 
so, we can feed output/\*/\*.tfrecord to run_pretraining script.

**Comment by zheolong** [comment]
@dsindex wiki_xx.txt is made from several wiki pages, ok, I've figured that.

So, what I want to say, is like below

![image](https://user-images.githubusercontent.com/5104916/50943178-4bd76c00-14c6-11e9-8ddb-b8a301bffa30.png)

**Comment by dsindex** [comment]
@zheolong

i got it. well, i think you’d better modify the wikiExtractor to achieve what you want.

**Comment by dsindex** [comment]
@zheolong 

there are two options we can choose. 

1. allow next sentence in a doc(a wiki page)
  - https://github.com/dsindex/bert/blob/master/preprocess.py
  ```
  lines = tokenize.sent_tokenize(line)
  for l in lines:
      print(l)
  ```
  - empty newline b/w docs and no newline in a doc
  - i think this is reasonable because even though the paragraphs are different, the topic is similar in that context. 

2. allow next sentence in a paragraph only
  - we can modify 'preprocess.py'
  ```
    lines = tokenize.sent_tokenize(line)
    for l in lines:
      print(l)
    print('')
  ```
  - empty newline b/w docs and b/w paragraphs
  - i think this is also reasonable.

i used the first method to create pretraining data.
but i did't compare which is better.

**Comment by zheolong** [comment]
@dsindex ok, got it

**Comment by dsindex** [comment]
@zheolong

vocab.txt.uncased cames from ‘glove.6B.’ official releases. 
and vocab.txt.cased are from ‘glove.840B.’.

note that these are for building word-based bert model.

and i also modified ‘tokenization.py’ for not using word-piece tokenizer. i only want to use basic tokenizer for my task.

you could download original vocab.txt from the official bert releases.

**Comment by zheolong** [comment]
@dsindex Do we need to set dupe_factor>1 for this big data set?

**Comment by dsindex** [comment]
@zheolong 

well, i don't know exactly. 
however, since the loss is calculated only for the masked tokens, the tf record data for training would be smaller than we expected if we set 'dupe_factor=1'.

**Comment by zheolong** [comment]
@dsindex what's ur machine and gpu configuration for pre-trainning？

**Comment by dsindex** [comment]
@dsindex 

i used this script for pre-training(BERT base, uncased).
https://github.com/dsindex/bert/blob/master/run_pretraining.bash
on Tesla V100(1 GPU, 8 CPU, 32 GB)

it takes 3 days long for 1M steps.

however, in case we are using word-piece tokenizer, the vocabulary should be different from mine(from glove.6B). so, the training time might be different(longer?).

**Comment by zheolong** [comment]
@dsindex With Wiki data? 3 days is really short. Google use four days on 4 to 16 Cloud TPUs

**Comment by dsindex** [comment]
@zheolong

yes, but i used word-based vocab and 1M steps and i don’t know what is the final steps they used for training.

i guess that word-piece vocab takes long time or they used huge training steps.

**Comment by zheolong** [comment]
@dsindex We have our own gpu cluster, so we prefer to use gpu

**Comment by zheolong** [comment]
@dsindex I've trained the model for 5.5 days, with (wiki + bookCorpus) dataset, my vocab is google public uncased vocab in ```uncased_L-12_H-768_A-12```, the vocab size is ```30522```. The loss and accuracy are as below.

```
INFO:tensorflow:***** Eval results *****
INFO:tensorflow:  global_step = 1000000
INFO:tensorflow:  loss = 1.7235886
INFO:tensorflow:  masked_lm_accuracy = 0.6630389
INFO:tensorflow:  masked_lm_loss = 1.6173733
INFO:tensorflow:  next_sentence_accuracy = 0.955
INFO:tensorflow:  next_sentence_loss = 0.1037216
```

The machine is with 2 Tesla P100 gpu, 16GB memory with each. Do u think it's normal?

**Comment by dsindex** [comment]
@zheolong

yes~ that is normal result.
but for getting a good model, you need to run more steps.
the ‘loss’ below 1.0 is good enough.

**Comment by micmelesse** [comment]
@dsindex Have you consider creating a PR for your script.

**Comment by eric-haibin-lin** [comment]
@zheolong that's really fast. did you use max-seq-length = 128 or 512? did you train from scratch, or initialized the model with the pre-trained weights?

**Comment by zheolong** [comment]
@eric-haibin-lin max-seq-length = 128, from scratch

**Comment by eric-haibin-lin** [comment]
Thanks for sharing @zheolong

**Comment by zheolong** [comment]
@dsindex 

I've found special commas in wiki dataset, its hex is '0xC2 0xA0', just like below. Is there any need to change them to normal '0x20'?

1. The original text

<img width="1431" alt="2019-02-20 3 56 25" src="https://user-images.githubusercontent.com/5104916/53139114-910db400-35c3-11e9-8a4f-6d81c1453abf.png">

2. The hex
￼
![10_34_43__02_21_2019](https://user-images.githubusercontent.com/5104916/53139389-738d1a00-35c4-11e9-9fbf-a71ae5abfe7f.jpg)

**Comment by zheolong** [comment]
Another problem is with bookcorpus dataset, I've found chinese comma like below, is that normal? 

![10_55_27__02_21_2019](https://user-images.githubusercontent.com/5104916/53140449-a4bb1980-35c7-11e9-8c52-a166560ac660.jpg)

**Comment by dsindex** [comment]
@zheolong

just my thought.

it will be a little bit better.
but, since we have a large corpus, the small change will not be considerable.

**Comment by micmelesse** [comment]
@dsindex Do you plan to create a PR adding sentence split.

**Comment by dsindex** [comment]
@micmelesse  no, i don't ^^;

**Comment by Frank1993** [comment]
> @dsindex I've trained the model for 5.5 days, with (wiki + bookCorpus) dataset, my vocab is google public uncased vocab in `uncased_L-12_H-768_A-12`, the vocab size is `30522`. The loss and accuracy are as below.
> 
> ```
> INFO:tensorflow:***** Eval results *****
> INFO:tensorflow:  global_step = 1000000
> INFO:tensorflow:  loss = 1.7235886
> INFO:tensorflow:  masked_lm_accuracy = 0.6630389
> INFO:tensorflow:  masked_lm_loss = 1.6173733
> INFO:tensorflow:  next_sentence_accuracy = 0.955
> INFO:tensorflow:  next_sentence_loss = 0.1037216
> ```
> 
> The machine is with 2 Tesla P100 gpu, 16GB memory with each. Is the result normal?

@zheolong, we also tried max sequence length as 128, it's very fast. Have you tried max sequence length to 512? in my experiemnts, it's really slow, as the batch size can only be 8 on a signle p100. We need lots of GPU to scale the batch.

Any one else tried shared memory of multiple GPU to suport larger batch size?

**Comment by 311dada** [comment]
> @dsindex I've trained the model for 5.5 days, with (wiki + bookCorpus) dataset, my vocab is google public uncased vocab in `uncased_L-12_H-768_A-12`, the vocab size is `30522`. The loss and accuracy are as below.
> 
> ```
> INFO:tensorflow:***** Eval results *****
> INFO:tensorflow:  global_step = 1000000
> INFO:tensorflow:  loss = 1.7235886
> INFO:tensorflow:  masked_lm_accuracy = 0.6630389
> INFO:tensorflow:  masked_lm_loss = 1.6173733
> INFO:tensorflow:  next_sentence_accuracy = 0.955
> INFO:tensorflow:  next_sentence_loss = 0.1037216
> ```
> 
> The machine is with 2 Tesla P100 gpu, 16GB memory with each. Is the result normal?

So do you deal with headers in documents?
